src/contrib/dependencyinjection/Akka.DependencyInjection/ServiceProvider.cs

```diff
@@ -0,0 +1,143 @@
+        /// such as a database connection, and disposing that scope once the operation is complete.
+        ///
+        /// Actors are not MVC Controllers. Actors can live forever, have the ability to restart, and are often stateful.
+        /// Be mindful of this as you use this feature or bad things will happen. Akka.NET does not magically manage scopes
+        /// for you.
+        /// </remarks>
+        public IServiceProvider Provider { get; }
+
+        public static ServiceProvider For(ActorSystem actorSystem)
+        {
+            return actorSystem.WithExtension<ServiceProvider, ServiceProviderExtension>();
+        }
+
+        /// <summary>
+        /// Uses a delegate to dynamically instantiate an actor where some of the constructor arguments are populated via dependency injection
+        /// and others are not.
+        /// </summary>
+        /// <remarks>
+        /// YOU ARE RESPONSIBLE FOR MANAGING THE LIFECYCLE OF YOUR OWN DEPENDENCIES. AKKA.NET WILL NOT ATTEMPT TO DO IT FOR YOU.
+        /// </remarks>
+        /// <typeparam name="T">The type of actor to instantiate.</typeparam>
+        /// <param name="producer">The delegate used to create a new instance of your actor type.</param>
+        /// <returns>A new <see cref="Props"/> instance which uses DI internally.</returns>
+        public Props Props<T>(Func<IServiceProvider, T> producer) where T : ActorBase
+        {
+            return new ServiceProviderProps<T>(producer, Provider);
+        }
+    }
+
+    /// <summary>
+    /// INTERNAL API
+    /// </summary>
+    public sealed class ServiceProviderExtension : ExtensionIdProvider<ServiceProvider>
+    {
+        public override ServiceProvider CreateExtension(ExtendedActorSystem system)
+        {
+            var setup = system.Settings.Setup.Get<ServiceProviderSetup>();
+            if (!setup.HasValue)
+            {
+                var exception = new ConfigurationException("Unable to find [ServiceProviderSetup] included in ActorSystem settings." +
+                                                           " Please specify one before attempting to use dependency injection inside Akka.NET.");
+                system.EventStream.Publish(new Error(exception, "Akka.DependencyInjection", typeof(ServiceProviderExtension), exception.Message));
+                throw exception;
+            }
+
+            return new ServiceProvider(setup.Value.ServiceProvider);
+        }
+    }
+
+    /// <summary>
+    /// This class represents a specialized <see cref="Akka.Actor.Props"/> that uses delegate invocation
+    /// to create new actor instances, rather than a traditional <see cref="System.Activator"/>.
+    ///
+    /// Relies on having an active <see cref="IServiceProvider"/> implementation available
+    /// </summary>
+    /// <typeparam name="TActor">The type of the actor to create.</typeparam>
+    internal class ServiceProviderProps<TActor> : Props where TActor : ActorBase
```

Comment:
So this class, the `ServiceProviderProps<TActor>`, is where the real magic happens. This class takes a delegate of type `Func<IServiceProvider, TActor>` and invokes that in order to create a new actor instance.

That function will be re-run each time the actor is restarted or each time a new actor is created using the same `Props` - so it's very easy to do things like:

1. Create a new `IServiceScope` and pass that into the actor, so the actor can manage its own lifecycle dependencies upon start / restart;
2. Manually resolve a dependency using `IServiceProvider.GetRequiredService<T>` and pass that directly into the actors' constructor - nice to have in the event that your actor doesn't need any scoped dependencies.